src/caffe/layers/conv_layer.cu

```diff
@@ -4,24 +4,170 @@
 
 namespace caffe {
 
+template <typename Dtype>
+__global__ void spatial_conv_fwd_kernel(const int count, const int channels, const int output_height, const int output_width,
+            const int input_height, const int input_width, 
+            const int kernel_height, const int kernel_width,
+            const int stride_y, const int stride_x, const int pad_y, const int pad_x,
+            const Dtype *input_data, const Dtype *weight_data, Dtype *output_data) { //, const Dtype *bias_data
+  CUDA_KERNEL_LOOP(index, count) {
+      output_data += index;
+      int ow = index % output_width;
+      index /= output_width;
+      int oh = index % output_height;
+      index /= output_height;
+      int c = index % channels;
+      int n = index / channels;
+
+      int iw = ow * stride_x - pad_x;
+      int ih = oh * stride_y - pad_y;
+
+      input_data += ((n * channels + c) * input_height + ih) * input_width + iw;
+      weight_data += c * kernel_width * kernel_height;
+
+      Dtype v = 0;
+      for (int kh = 0; kh < kernel_height; kh++) {
+          if (ih + kh >= 0 && ih + kh < input_height) {
+              for (int kw = 0; kw < kernel_width; kw++) {
+                  if (iw + kw >= 0 && iw + kw < input_width) {
+                      v += input_data[kw] * weight_data[kw];
+                  }
+              }
+          }
+          input_data += input_width;
+          weight_data += kernel_width;
+      }
+
+      // if (bias_data) {
+      //     v += bias_data[c];
+      // }
+
+      *output_data = v;
+  }
+}
+
 template <typename Dtype>
 void ConvolutionLayer<Dtype>::Forward_gpu(const vector<Blob<Dtype>*>& bottom,
       const vector<Blob<Dtype>*>& top) {
   const Dtype* weight = this->blobs_[0]->gpu_data();
   for (int i = 0; i < bottom.size(); ++i) {
     const Dtype* bottom_data = bottom[i]->gpu_data();
     Dtype* top_data = top[i]->mutable_gpu_data();
-    for (int n = 0; n < this->num_; ++n) {
-      this->forward_gpu_gemm(bottom_data + n * this->bottom_dim_, weight,
-          top_data + n * this->top_dim_);
-      if (this->bias_term_) {
-        const Dtype* bias = this->blobs_[1]->gpu_data();
-        this->forward_gpu_bias(top_data + n * this->top_dim_, bias);
+    if (this->group_ == this->channels_){
```

Comment:
In the description, you mention a case when group==num_output, yet here you compare it to channels instead. Is this correct?  
Also, you don't really need `this->` when accessing attributes.